editor/contexts/ThemeProvider.tsx

```diff
@@ -0,0 +1,40 @@
+"use client";
+import { useEffect, useState, createContext, useContext } from "react";
+import { darkModeLocalStorageKey } from "@/components/DarkModeToggle";
+
+export const ThemeContext = createContext<string | null>(null);
+
+/**
+ * Global context provider that returns whether the app is in dark or light mode
+ */
+export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({
+  children,
+}) => {
+  const [themeMode, setThemeMode] = useState(
+    localStorage.getItem(darkModeLocalStorageKey)
+  );
```

Comment:
It might be good to copy the typing (and var name?) we use in the `DarkModeToggle`, since it is tighter:

```typescript
  const [appColorMode, setAppColorMode] = useState<"dark" | "light">(
    localStorage.getItem(darkModeLocalStorageKey) === "dark" ? "dark" : "light"
  );
```

Or instead of `themeMode`, short and sweet: `theme` ?

Reply:
Hmm, now I'm having doubts about including this `localStorage.getItem` call in the `DarkModeToggle`. It feels a bit off that we are pulling from `localStorage` to check for dark mode in two different places.

Maybe `DarkModeToggle` should use the value from the new context instead? I suppose that is complicated by the toggle itself using this state.

Going a step further, maybe the useEffect code in `DarkModeToggle` could be centralized in the context, and `useTheme` provides that as well, so you can do this

```typescript
const [themeMode, setThemeMode] = useTheme()
```

Am I overthinking this?

Reply:
Agh, I just noticed I get a build error locally that the app seems to recover from. 

```
⨯ contexts/ThemeProvider.tsx (14:5) @ localStorage
 ⨯ ReferenceError: localStorage is not defined
    at ThemeProvider (./contexts/ThemeProvider.tsx:19:87)
digest: "3899942164"
  12 | }) => {
  13 |   const [themeMode, setThemeMode] = useState(
> 14 |     localStorage.getItem(darkModeLocalStorageKey)
     |     ^
  15 |   );
```

It might be that the initial `localStorage` interaction needs to be inside `useEffect`? 😐

Reply:
Okay so I was able to move all of the state setting and getting from `DarkModeToggle` along with the contents of its `useEffect` that sets the local storage and DOM into the `AppThemeProvider` and it worked beautifully.  This allowed me to actually cut down on some logic and remove the first `useEffect` that added the event listener bc all of the theme state context is contained here and is global so we dont even need that custom event listener stuff anymore :smile:

Reply:
Oh i am also now getting that build error... lemme see whats up

Reply:
Okay yeah, it was a server side rendering problem, we were trying to access localStorage during SSR. as you said, had to employ a `useEffect` to update the state w the local `localStorage` val